### DISCUSSION TITLE
Add tire pressure sensor device class

### DISCUSSION THREAD

**Discussion by bdr99** [home-assistant/architecture#532] [Entity Models] [answered]
I recently submitted a PR (https://github.com/home-assistant/core/pull/47951) to add support for converting between bar and other pressure units, but @MartinHjelmare suggested that I submit an architecture issue first.

Home Assistant does not currently have an appropriate device class for tire pressure sensors. There is a device class for pressure, but it seems that this device class was intended for atmospheric pressure, and it is currently mostly used in weather and climate integrations. According to the [documentation](https://developers.home-assistant.io/docs/core/entity/sensor/#available-device-classes), the pressure device class mandates units of hPa and mbar, whereas tire pressure is measured using psi in the US, and bar in Europe.

I propose that we take one of the following actions:

1. Expand the existing pressure device class to allow additional units other than hPa and mbar, so that it can be used for tire pressure sensors in addition to atmospheric pressure sensors.
2. Create a new device class for tire pressure sensors, and consider renaming the existing `pressure` device class to `atmospheric_pressure`.

Several of our Car integrations implement tire pressure sensors. The [Volvo On Call](https://www.home-assistant.io/integrations/volvooncall/) integration has supported tire pressure for a long time, and the new [Subaru](https://www.home-assistant.io/integrations/subaru/) and [Mazda](https://www.home-assistant.io/integrations/mazda/) integrations, added in 2021.03, also support tire pressure sensors. These integrations would benefit from having a more appropriate device class to use for their tire pressure sensor entities.

**Comment by frenck** [comment]
PSI/Bar units are not solely used for car tires, but are probably very useful for boilers as well. Opentherm is an example of such an integration that could support that.

The OpenTherm GW integration currently defines:

```py
# Pressure units
PRESSURE_PA: str = "Pa"
PRESSURE_HPA: str = "hPa"
PRESSURE_BAR: str = "bar"
PRESSURE_MBAR: str = "mbar"
PRESSURE_INHG: str = "inHg"
PRESSURE_PSI: str = "psi"
```

I think I would not add a tire pressure entity/device class, mainly it is not the solve measurement factor for a tire (mile/km, tread wear)..

**Reply by bdr99** [reply]
I agree that option 2 is not ideal here. We already have a device class for pressure, so I think it makes sense to just expand the list of allowable units for that device class (option 1).

Also, in `homeassistant.const` we already have [constants](https://github.com/home-assistant/core/blob/dev/homeassistant/const.py#L429) for 6 different pressure units: Pa, hPa, bar, mbar, inHg, and psi (the same ones as in the OpenTherm integration). I don't see any reason that those units shouldn't be allowed for the pressure device class if they are already declared as constants and used in various places in HA.

**Reply by MartinHjelmare** [reply]
Maybe add a device class `DEVICE_CLASS_PRESSURE_VESSEL` with unit bar, localize this in the Entity / SensorEntity class?

It should represent pressure measurements of all types of containers, tires, tanks, tubes etc, that have a different pressure than atmosphere pressure.

**Reply by bdr99** [reply]
@MartinHjelmare This sounds like a good idea to me. There's definitely a clear difference between devices that measure atmospheric pressure and devices that measure pressure inside a container.

**Comment by MartinHjelmare** [comment]
My general comment about device classes and units is that I'd like each device class to have a specific unit. Preferably the base entity should then localize that unit (imperial / metric) before it is written in the state machine. We don't do this consistently yet, only for temperature, but that's what I would want to aim for.

**Reply by bdr99** [reply]
I certainly agree that this would be the ideal scenario. However, since we aren't there yet, I think we still need an interim solution.

Currently in the Mazda integration, I am unable to localize my PSI tire pressure units to bar, because of the lack of support for bar in `homeassistant.util.pressure`. This is causing an undesirable experience for users (see https://github.com/home-assistant/home-assistant.io/issues/16999). Do you think it would be acceptable for the tire pressure sensors to continue to use the generic device class for now, at least until we reach the point where the base entity is capable of doing the unit conversions automatically?

**Comment by emontnemery** [comment]
Integrations which would benefit from this still seem to be:
`mazda`, `subaru`, `volvooncall`

**Reply by bdr99** [reply]
Also, if the device class were expanded to encompass all pressure vessels instead of just tires, like @MartinHjelmare suggested above, then the following integrations would also benefit:
`flo` - water pressure
`oncue` - engine oil pressure
`wolflink` - water pressure

I know that at least in the US, we use PSI in all of these cases, so the unit would be the same for all pressure vessels including car tires. I think the same generalization can be made in other regions where bar is used for tire pressure, but I'm not sure of this.

**Reply by epenet** [reply]
We recently added a generic `Pressure` device class for these cases.
I think liquid pressures should not be confused with type pressures.

I do think that tyre pressure merits it's own device class, and you could use https://github.com/home-assistant/core/pull/83455 as a sample implementation.

**Reply by emontnemery** [reply]
You and Martin are right, `kPa`, or possibly `bar`, would work also for sensors measuring oil and water pressure 👍 
Since we now allow users to customize the unit of sensors from the UI, it's no longer crucial if the default metric unit is `bar` or `kPa`.

**Reply by bdr99** [reply]
I can work on a PR for this, but I just want to ask for your opinions on a few things first. @emontnemery , @epenet, @MartinHjelmare 

1. What should the new device class(es) be named? If we are going with a single device class for both tire pressure and pressure vessels, then maybe something like `fluid_pressure` (since air is also considered a fluid)? Or if we are going with two device classes, then maybe `liquid_pressure` and `gas_pressure`? It might be a bit confusing to have both `atmospheric_pressure` and `gas_pressure`, since the ambient air is also a gas, it's just not in a pressurized vessel. What are your thoughts?

2. It seems that the new device class(es) for fluid pressure, in combination with the `atmospheric_pressure` device class being added in https://github.com/home-assistant/core/pull/83455, would cover all cases where the `pressure` device class is currently being used today. So are we planning to move all integrations currently using `pressure` to either `atmospheric_pressure` or the new fluid pressure, and eventually deprecate and remove `pressure`?

**Reply by epenet** [reply]
If there are only 6 integrations (3 tire, 2 liquid and 1 engine oil) then maybe the existing generic `pressure` is sufficient, now that all pressure units are allowed?
It allows users to select their preferred icon and unit, even if it doesn't (yet) do automatic conversion.
Maybe we could also add automatic conversions `psi <-> bar` and leave other units left untouched (see https://github.com/home-assistant/core/pull/83525).

Regarding `atmospheric_pressure`, I think we have so many integrations that it does make sense to "isolate" it from the generic `pressure` device class, with units standardized to `inHg` for USCS and `hPa` for metric

**Reply by MartinHjelmare** [reply]
Yeah, if we break out atmospheric pressure the need for another specific pressure class becomes smaller. I'm in favor of holding off on adding this other pressure device class for now. If more integrations are added that need it in the future we can revisit.

**Reply by bdr99** [reply]
I agree, but should the generic `pressure` device class be renamed to clarify that it is not to be used for atmospheric pressure? Or should this at least be documented somewhere? As it stands, we will have both `pressure` and `atmospheric_pressure`, which are not mutually exclusive, and I can imagine that some users may be a bit confused by that.

**Reply by epenet** [reply]
I think `pressure` should not be renamed. It stays generic and can be used by any component that uses any pressure unit.

It will not be wrong to set the device class to the generic `pressure` for weather entities.
It just won't be optimum because it will not have automatic conversion enabled.

Maybe adjusting the docs and docstring could be done... you can add review comments on https://github.com/home-assistant/core/pull/83525 and the linked PRs

**Comment by emontnemery** [comment, marked as answer]
We will break out an atmospheric pressure device class which makes the need for another specific pressure class smaller. If more integrations are added that need it in the future, we can revisit.
